# Introduction 
As a data analyst, your job is to analyze data to extract valuable information and make decisions based on it. This involves different stages, such as data overview, preprocessing and hypothesis testing.

Whenever we do research, we need to formulate hypotheses that we can then test. Sometimes we accept these hypotheses; sometimes we reject them. To make the right decisions, a company must be able to understand whether it is making the right assumptions.

In this project, you will compare the music preferences of the cities of Springfield and Shelbyville. You will study real online music streaming data to test the hypotheses below and compare the behavior of male and female users in these two cities.

## Goal 
Tests three hypotheses:

+ Male and female user activity differs by day of the week and depending on the city.
+ On Monday mornings, people in Springfield and Shelbyville listen to different genres. The same is true on Friday nights.
+ Springfield and Shelbyville listeners have different preferences. In Springfield they prefer pop, while in Shelbyville more people like rap.

## Stages
The user behavior data is stored in the file /datasets/music_project_en.csv. There is no information about the quality of the data, so you will need to examine it before testing hypotheses.

First, you will assess the quality of the data and see if the problems are significant. Then, during data preprocessing, you will consider the most critical problems.

Your project will consist of three stages:

+ Data description
+ Data preprocessing
+ Hypothesis testing


 

# Stage 1 - Data Description

In [64]:
# Import pandas
import pandas as pd

In [65]:
# Load the dataset 
df = pd.read_csv('music_project_en.csv')
df 

userID                              Track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina               NaN   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre       City        time        Day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfield  14:07:09     Friday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
4           dance  Springfield  08:34:34     Monday  
...           ...          ...       ...        ...  
65074         rnb  Springfield  13:32:28  Wednesday  
65075         hip  Shelbyville  10:00:00     Monday  
65076  industrial  Springfield  20:09:26     Friday  
65077        rock  Springfield  21:43:59     Friday  
65078     country  Springfield  21:59:46     Friday  

[65079 rows x 7 columns]

In [66]:
# Load the first 10 rows 
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

In [67]:
# Get the general info about the dataset 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


## Observations 

### Is there enough data to provide answers to our three hypotheses or do we need more information?
Hypothesis 1 (differences by day of the week and city): Differences in user activity by day of the week and city can be explored with the available data. However, the amount of missing data in some columns, such as 'Track,' 'artist,' and 'genre,' could limit the accuracy of the conclusions. The amount of non-zero data in these columns will be crucial in evaluating this hypothesis.

Hypothesis 2 (differences on Monday mornings and Friday nights): To test this hypothesis, it is necessary to ensure that the data contain information on the exact time each song was played ('time' column). If there is an adequate distribution of data over time, you will be able to assess whether there are significant differences on Monday mornings and Friday nights.

Hypothesis 3 (music genre preferences in Springfield and Shelbyville): To evaluate this hypothesis, you need to examine the 'genre' column and the 'City' column to determine if there are significant differences in music genre preferences between the two cities. This is feasible if you have sufficient non-zero data in these columns.

In summary, although it is possible to perform analyses related to all three hypotheses with the available data, the accuracy and reliability of the conclusions will depend heavily on the quantity and quality of the non-null data in the relevant columns. It would be good to perform a detailed exploration of the data and consider addressing missing values before performing more in-depth analyses.

# Stage 2 - Data Pre - processing 


In [68]:
# Headings Type 
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


The columns are written in a mixed style with upper and lower case, which can cause problems or difficulties when working with them, problems such as misspellings. To avoid these possible problems it is better to work with the names completely in lowercase and remove all spaces from the name, for this we can use the functions 'lower.()' and 'strip()'.

## Getting the columns to lowercases 

Since the 'Index' object has no attribute 'lower' we must find another way to complete this task than just applying the function into the columns. It will be better if we take out the names and put them into an empty list so we can use the function and then just assign that list with all the names in lowercases as the columns of our data frame

In [69]:
# Create an empty list where we save all of the columns' names 
columns = []

# Iterate on each name from the list and apply 'lower()' 
for column in df.columns: 
    lowcase = column.lower()
    columns.append(lowcase)

# Assign the new names to the columns 
df.columns = columns

# Show the final result 
print(df.columns)

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


## Getting rid of the spaces in the names 

We use the same logic that we used in the last task

In [70]:
# Create an empty list to add the names without spaces
new_columns = []

# Iterate on each name to get rid of the spaces in the names
for column in df.columns: 
    strip = column.strip()
    new_columns.append(strip)

# Assign the new names to the columns of the data frame
df.columns = new_columns

# Show the result 
print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


We completed the task, now we don't have to worry if the names of the columns have uppercases in different postions or spaces at the beggining or end

## Missing Values 
There can be cases where the missing values can not affect the project but we still need to study them to see if they can affect us or not

In [71]:
# Find the missing values in the data frame 
print(df.isna().sum())

userid       0
track     1343
artist    7567
genre     1198
city         0
time         0
day          0
dtype: int64


Not all missing values affect the investigation. For example, missing values in track and artist are not crucial. You can simply replace them with default values such as the string 'unknown'.

But missing values in 'genre' may affect the comparison between Springfield and Shelbyville music preferences. In real life, it would be useful to know the reasons for missing data and try to recover them. But we don't have that opportunity in this project. So you will have to:

+ Fill in these missing values with a default value;
+ Evaluate how much the missing values might affect your computations;



In [72]:
# Replace the missing values with the string 'unknown'  
columns_to_replace = ['track','artist','genre']

# Create a loop to fill all the missing values at once 
for col in columns_to_replace:
    df[col].fillna('unknown', inplace=True)

# Show the results 
print(df.isna().sum())
display(df)

userid    0
track     0
artist    0
genre     0
city      0
time      0
day       0
dtype: int64


C:\Users\User\AppData\Local\Temp\ipykernel_22324\2664617334.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna('unknown', inplace=True)


userid                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina           unknown   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfield  14:07:09     Friday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
4           dance  Springfield  08:34:34     Monday  
...           ...          ...       ...        ...  
65074         rnb  Springfield  13:32:28  Wednesday  
65075         hip  Shelbyville  10:00:00     Monday  
65076  industrial  Springfield  20:09:26     Friday  
65077        rock  Springfield  21:43:59     Friday  
65078     country  Springfield  21:59:46     Friday  

[65079 rows x 7 columns]

## Duplicate Values 
Handling duplicate values differs from dealing with missing data because duplicates consistently compromise the accuracy of data analysis. Leaving duplicate records in a DataFrame can lead to biased statistical results by counting the same observation multiple times, ultimately affecting insights and decision-making. Therefore, identifying and removing duplicates is a crucial preprocessing step to ensure the integrity of the analysis.

In [73]:
# Find the duplicate values 
print(df.duplicated().sum())

3826


There are many duplicates in the data frame, so the best thing to do is just eliminating them. We use 'pandas' for that, using 'drop.duplicates()' we'll take care of the problem


In [74]:
# Get rid of the duplicates with 'drop_dupliicates()'    
df = df.drop_duplicates().reset_index(drop=True)

# Confirm that there are no duplicates left
print(df.duplicated().sum())  

0


Now we want to get rid of the implicit duplicates in the genre column. For example, the name of a genre can be written in several ways. Such errors can also affect the result.

To do this, let us first display a list of unique genre names, sorted in alphabetical order. To do this:

retrieve the desired column from the dataFrame;
call the method that will return all unique column values;
apply a sort method to your result.



In [75]:
# Show the unique values of 'genre' 
print(df['genre'].unique())

print()

# Show how many unique values are in the df 
print(f'There are {df['genre'].nunique()} unique genres in the data frame')

['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hip' 'jazz' 'postrock' 'latin'
 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental' 'rusrock'
 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'hiphop' 'drum' 'extrememetal' 'türkçe'
 'experimental' 'easy' 'metalcore' 'modern' 'argentinetango' 'old' 'swing'
 'breaks' 'eurofolk' 

Search the list to find implied duplicates of the hiphop genre. These may be misspelled names or alternate names for the same genre.

You will see the following implied duplicates:

hip, hop, hip-hop

To get rid of them, declare the replace_wrong_genres() function with two parameters:

+ wrong_genres= - the list of duplicates;
+ correct_genre= - the string with the correct value.

The function should correct the names in the 'genre' column of the df table, i.e. it replaces each value in the wrong_genres list with the value in correct_genre. It uses a for loop to iterate over the list of incorrect genres and replace them with the correct genre in the main list.

In [76]:
# Declare the function 'replace_wrong_genre()'
def replace_wrong_genre(wrong_genres, correct_genre): 
    # Replace the wrong genres with the correct one in the list
    for gen in wrong_genres:
        df['genre'].replace(wrong_genres, correct_genre, inplace=True)
    return f'Now there are {df['genre'].nunique()} unique genres in the data frame'

In [77]:
replace_wrong_genre(['hip','hop','hip-hop'], 'hiphop')

C:\Users\User\AppData\Local\Temp\ipykernel_22324\3029943233.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['genre'].replace(wrong_genres, correct_genre, inplace=True)


'Now there are 266 unique genres in the data frame'

At the beggining we has 269 unique genres in the data frame, incluiding the implied duplicates of the genre 'hiphop'. We took care of that problem and now we have 266 unique genres in the data frame and by that we can confirm that we updated all the 'hiphop' values to just one

# Stage 3: Hypothesis Testing 
## Hypothesis 1: Comparing user behavior in the two cities 

The first hypothesis states that there are differences in the way male and female users in Springfield and Shelbyville consume music. To test this, it uses data from three days of the week: Monday, Wednesday, and Friday.

Group the users by city.
+ Compare the number of songs each group played on Monday, Wednesday, and Friday.
+ Perform each calculation separately.

To evaluate the activity of users in each city, group the data by city and find the number of songs played in each group.


In [78]:
# Count the total reps by city
reps_by_city = df.groupby('city')['track'].count()
# Show the result
display(reps_by_city)

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

Now group the data by day of the week and find the number of songs played on Monday, Wednesday and Friday.

In [79]:
# Count the amount of total replays on each day
reps_by_day = df.groupby('day')['track'].count()
# Show the result
display(reps_by_day)

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

You already know how to count entries by grouping them by city or day. Now you need to write a function that can count entries by both criteria simultaneously.

Create the number_tracks() to calculate the number of songs played on a given day and city. The function must accept two parameters:

+ 'day': a day of the week to filter. For example, 'Monday'.
+ city: a city to filter. For example, 'Springfield'.

Inside the function, you will apply consecutive filtering with logical indexing.

First filter the data by day and then filter the resulting table by city.

After filtering the data by two criteria, count the number of values in the 'user_id' column in the resulting table. This count represents the number of entries you are looking for. Save the result in a new variable and return it from the function.

In [80]:
# <creating the function number_tracks()>
# We will declare the function with two parameters: day=, city=.
# Let the track_list variable store the df rows in which
# the value of the column name 'day' is equal to the parameter day= and, at the same time,
# the value of the column name 'city' is equal to the parameter city= (apply consecutive filtering with logical indexing).
# with logical indexing).
# let the track_list_count variable store the number of values of the column 'user_id' in track_list
# (equal to the number of rows in track_list after filtering twice).
# allows the function to return a number: the value of track_list_count.

# function counts the tracks played on a certain day and city.
# first retrieves the rows of the desired day from the table,
# then it filters the rows of the desired city from the result,
# then it finds the number of tracks in the filtered table,
# and returns that number.
# To see what it returns, wrap the function call in print().

def number_tracks(day, city): 
    track_list = df[(df['day'] == day) & (df['city'] == city)]

    track_list_count = track_list['userid'].count()

    return track_list_count

### Total reps by day for Springfield

In [85]:
# Total reps on Monday 
print(f'The total songs played on Monday in Springfield are: {number_tracks('Monday','Springfield')}')
# Total reps on Wednesday
print(f'The total songs played on Wednesday in Springfield are: {number_tracks("Wednesday","Springfield")}')
# Total reps on Friday 
print(f'The total songs played on Friday in Springfield are: {number_tracks("Friday","Springfield")}')

The total songs played on Monday in Springfield are: 15740
The total songs played on Wednesday in Springfield are: 11056
The total songs played on Friday in Springfield are: 15945


### Total reps by day on Shelbyville


In [86]:
# Total reps on Monday 
print(f'The total songs played on Monday in Shelbyville are: {number_tracks("Monday","Shelbyville")}')
# Total reps on Wednesday
print(f'The total songs played on Wednesday in Shelbyville are: {number_tracks("Wednesday","Shelbyville")}')
# Total reps on Friday 
print(f'The total songs played on Friday in Shelbyville are: {number_tracks("Friday","Shelbyville")}')

The total songs played on Monday in Shelbyville are: 5614
The total songs played on Wednesday in Shelbyville are: 7003
The total songs played on Friday in Shelbyville are: 5895


## Result Hypothesis 1 


In [88]:
table = pd.DataFrame({
    'city' : ['Springfield', 'Shelbyville'],
    'Monday': [15740, 5614],
    'Wednesday' : [11056, 7003],
    'Friday' : [15945, 5895]
})

display(table)

city  Monday  Wednesday  Friday
0  Springfield   15740      11056   15945
1  Shelbyville    5614       7003    5895

### Observing General Trends

Springfield consistently has higher music consumption than Shelbyville on all three days.

In Springfield, music consumption is highest on Friday (15,945), slightly higher than Monday (15,740), and lowest on Wednesday (11,056).

In Shelbyville, Wednesday has the highest number of song plays (7,003), while Monday (5,614) and Friday (5,895) have similar, lower values.

### Identifying Differences Between Cities
Springfield has significantly more song reproductions than Shelbyville on all three days.

+ On Monday, Springfield plays ~2.8 times more songs than Shelbyville.
+ On Wednesday, Springfield plays ~1.58 times more songs than Shelbyville.
+ On Friday, Springfield plays ~2.7 times more songs than Shelbyville.

The pattern of weekly fluctuations is different:

In Springfield, Wednesday is the lowest day for song consumption.

In Shelbyville, Monday and Friday are the lowest, while Wednesday is the highest.

### Possible Interpretations
The data suggests that music consumption behavior differs between cities, supporting the hypothesis.

Possible reasons:

Demographics: Different age groups or lifestyles might influence music consumption patterns.

Work or school schedules: Shelbyville may have a midweek peak in music consumption, while Springfield shows a more traditional Friday peak.

Social or cultural factors: Local events, commuting habits, or access to music streaming services could influence the numbers.


### Conclusion

Based on the provided data, there are clear differences in music consumption between Springfield and Shelbyville. 

## Hypothesis 2: music at the beginning and end of the week 
According to the second hypothesis, on Monday morning and Friday night, Springfield citizens listen to different genres than those enjoyed by Shelbyville users.

Create two tables with the names provided in the following two code blocks:

+ For Springfield - spr_general
+ For Shelbyville - shel_general